onnxruntime/core/platform/windows/device_discovery.cc

```diff
@@ -362,89 +362,109 @@ std::unordered_map<uint64_t, DeviceInfo> GetDeviceInfoD3D12() {
   return device_info;
 }
 
+typedef HRESULT(WINAPI* PFN_DXCoreCreateAdapterFactory)(REFIID riid, void** ppvFactory);
+
 // returns LUID to DeviceInfo
 std::unordered_map<uint64_t, DeviceInfo> GetDeviceInfoDxcore() {
   std::unordered_map<uint64_t, DeviceInfo> device_info;
 
-  // Get all GPUs and NPUs by querying WDDM/MCDM.
-  wil::com_ptr<IDXCoreAdapterFactory> adapterFactory;
-  if (FAILED(DXCoreCreateAdapterFactory(IID_PPV_ARGS(&adapterFactory)))) {
+  // Load dxcore.dll. We do this manually so there's not a hard dependency on dxcore which is newer.
+  wil::unique_hmodule dxcore_lib{LoadLibraryExW(L"dxcore.dll", nullptr, LOAD_LIBRARY_SEARCH_SYSTEM32)};
+  if (!dxcore_lib) {
+    LOGS_DEFAULT(INFO) << "Failed to load dxcore.dll. Expected on older Windows version that do not support dxcore.";
     return device_info;
   }
 
-  // NOTE: These GUIDs requires a newer Windows SDK than what we target by default.
-  // They were added in 10.0.22621.0 as part of DirectXCore API
-  // To workaround this we define a local copy of the values. On an older Windows machine they won't match anything.
-  static const GUID local_DXCORE_ADAPTER_ATTRIBUTE_D3D12_GENERIC_ML = {0xb71b0d41, 0x1088, 0x422f, 0xa2, 0x7c, 0x2, 0x50, 0xb7, 0xd3, 0xa9, 0x88};
-  static const GUID local_DXCORE_HARDWARE_TYPE_ATTRIBUTE_NPU = {0xd46140c4, 0xadd7, 0x451b, 0x9e, 0x56, 0x6, 0xfe, 0x8c, 0x3b, 0x58, 0xed};
-
-  // Look for devices that expose compute engines
-  std::vector<const GUID*> allowedAttributes;
-  allowedAttributes.push_back(&DXCORE_ADAPTER_ATTRIBUTE_D3D12_CORE_COMPUTE);
-  allowedAttributes.push_back(&local_DXCORE_ADAPTER_ATTRIBUTE_D3D12_GENERIC_ML);
-  allowedAttributes.push_back(&local_DXCORE_HARDWARE_TYPE_ATTRIBUTE_NPU);
-
-  // These attributes are not OR'd.  Have to query one at a time to get a full view.
-  for (const auto& hwAttribute : allowedAttributes) {
-    wil::com_ptr<IDXCoreAdapterList> adapterList;
-    if (FAILED(adapterFactory->CreateAdapterList(1, hwAttribute, IID_PPV_ARGS(&adapterList)))) {
-      continue;
+  do {
+    auto pfnDXCoreCreateAdapterFactory = reinterpret_cast<PFN_DXCoreCreateAdapterFactory>(
+        GetProcAddress(dxcore_lib.get(), "DXCoreCreateAdapterFactory"));
+
+    if (!pfnDXCoreCreateAdapterFactory) {
+      // this isn't expected to fail so ERROR not WARNING
+      LOGS_DEFAULT(ERROR) << "Failed to get DXCoreCreateAdapterFactory function address.";
+      break;
```

Comment:
maybe simply `return device_info`, now that we don't need to explicitly call FreeLibrary at the end? and the do/while(false) doesn't seem to add much now.